aesd-char-driver/main.c

```diff
@@ -14,112 +11,309 @@
 #include <linux/module.h>
 #include <linux/init.h>
 #include <linux/printk.h>
-#include <linux/types.h>
+#include <linux/types.h> //size_t    
 #include <linux/cdev.h>
-#include <linux/slab.h>
-#include <linux/mutex.h>
+#include <linux/kernel.h> //containerof
+#include <linux/slab.h> //kmalloc
+#include <linux/uaccess.h>	/* copy_*_user */
 #include <linux/fs.h> // file_operations
+
 #include "aesdchar.h"
+#include "aesd_ioctl.h"
 int aesd_major =   0; // use dynamic major
 int aesd_minor =   0;
 
+#define DELIMITER           ('\n')
+
 MODULE_AUTHOR("abaa3023"); /** TODO: fill in your name **/
 MODULE_LICENSE("Dual BSD/GPL");
 
 struct aesd_dev aesd_device;
 
+static int allocate_memory(struct working_buffer *ptr, int size)
+{
+    int retval = 0;
+    char *temp_ptr;
+    if(ptr->size == 0)
+    {
+        ptr->buffer = kmalloc(size,GFP_KERNEL);
+        if(!ptr->buffer)
+        {
+            retval = -ENOMEM;  
+        }
+    }
+    else
+    {
+        temp_ptr = krealloc(ptr->buffer,ptr->size + size,GFP_KERNEL);
+        if(!temp_ptr)
+        {
+            kfree(ptr->buffer);
+            retval = -ENOMEM; 
+        }
+        else
+        {
+            ptr->buffer = temp_ptr;
+        }
+    }
+    return retval;
+}
+
+
+#ifdef __KERNEL__
```

Comment:
@abaa3023 I don't think this should be needed here... this file won't be compiled unless you are building the driver.